File: src/test/kotlin/model/ForestTest.kt
package model

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

internal class ForestTest {
    private lateinit var forest: Forest

    @BeforeEach
    fun setUp() {
        forest = Forest()
    }

    @Test
    fun `test makeSet creates a new set`() {
        forest.makeSet("someElement")

        assertEquals(forest.predecessor.size, 1)
        assertEquals(forest.predecessor["someElement"], "someElement")

        assertEquals(forest.rank.size, 1)
        assertEquals(forest.rank["someElement"], 0)
    }

    @Test
    fun `test union merges two sets and updates root node`() {
        forest.makeSet("someElement")
        forest.makeSet("anotherElement")

        forest.union("someElement", "anotherElement")

        assertEquals(forest.predecessor["someElement"], "anotherElement")
        assertEquals(forest.predecessor["anotherElement"], "anotherElement")
    }

    @Test
    fun `test union merges two sets and updates ranks`() {
        forest.makeSet("someElement")
        forest.makeSet("anotherElement")

        forest.union("someElement", "anotherElement")

        assertEquals(forest.rank["someElement"], 0)
        assertEquals(forest.rank["anotherElement"], 1)
    }

    @Test
    fun `test union by rank`() {
        forest.makeSet("1stElement")
        forest.makeSet("2ndElement")
        forest.makeSet("3rdElement")

        forest.union("2ndElement", "1stElement")

        /**
         * Instead of using 3rdElement as the new root node,
         * it uses 1stElement as the new root node,
         * because higher ranks will always win over dominantNode.
         */
        forest.union("1stElement", "3rdElement")

        assertEquals(forest.predecessor["1stElement"], "1stElement")
        assertEquals(forest.predecessor["2ndElement"], "1stElement")
        assertEquals(forest.predecessor["3rdElement"], "1stElement")
    }

    @Test
    fun `test union is always merging root nodes`() {
        forest.makeSet("1stElement")
        forest.makeSet("2ndElement")
        forest.makeSet("3rdElement")
        forest.makeSet("4thElement")

        assertEquals(forest.predecessor["1stElement"], "1stElement")
        assertEquals(forest.predecessor["2ndElement"], "2ndElement")
        assertEquals(forest.predecessor["3rdElement"], "3rdElement")
        assertEquals(forest.predecessor["4thElement"], "4thElement")

        forest.union("2ndElement", "1stElement")
        forest.union("4thElement", "3rdElement")

        assertEquals(forest.predecessor["1stElement"], "1stElement")
        assertEquals(forest.predecessor["2ndElement"], "1stElement")
        assertEquals(forest.predecessor["3rdElement"], "3rdElement")
        assertEquals(forest.predecessor["4thElement"], "3rdElement")

        forest.union("3rdElement", "1stElement")

        assertEquals(forest.predecessor["1stElement"], "1stElement")
        assertEquals(forest.predecessor["2ndElement"], "1stElement")
        assertEquals(forest.predecessor["3rdElement"], "1stElement")
        assertEquals(forest.predecessor["4thElement"], "3rdElement")
    }

    @Test
    fun `test findSet with path compression`() {
        forest.makeSet("1stElement")
        forest.makeSet("2ndElement")
        forest.makeSet("3rdElement")

        forest.union("2ndElement", "1stElement")

        assertEquals(forest.predecessor["3rdElement"], "3rdElement")

        forest.union("3rdElement", "2ndElement")

        assertEquals(forest.findSet("3rdElement"), "1stElement")
        assertEquals(forest.predecessor["3rdElement"], "1stElement")
    }

    @Test
    fun `test findSet fails for unknown element`() {
        assertThrows<NullPointerException> {
            forest.findSet("nonExistingElement")
        }
    }

    @Test
    fun `test toString provides correct representation`() {
        forest.makeSet("someElement")
        forest.makeSet("anotherElement")

        forest.union("someElement", "anotherElement")
        val expected = "Forest(predecessor={someElement=anotherElement, anotherElement=anotherElement}, rank={someElement=0, anotherElement=1})"
        assertEquals(expected, forest.toString())
    }
}

File: src/__tests__/model/forest.test.ts
import { Forest } from '../../lib/model/forest';

describe('Forest', () => {
  let forest: Forest;

  beforeEach(() => {
    forest = new Forest();
  });

  test('test makeSet creates a new set', () => {
    forest.makeSet('someElement');

    expect(Object.keys(forest.predecessor).length).toBe(1);
    expect(forest.predecessor['someElement']).toBe('someElement');

    expect(Object.keys(forest.rank).length).toBe(1);
    expect(forest.rank['someElement']).toBe(0);
  });

  test('test union merges two sets and updates root node', () => {
    forest.makeSet('someElement');
    forest.makeSet('anotherElement');

    forest.union('someElement', 'anotherElement');

    expect(forest.predecessor['someElement']).toBe('anotherElement');
    expect(forest.predecessor['anotherElement']).toBe('anotherElement');
  });

  test('test union merges two sets and updates ranks', () => {
    forest.makeSet('someElement');
    forest.makeSet('anotherElement');

    forest.union('someElement', 'anotherElement');

    expect(forest.rank['someElement']).toBe(0);
    expect(forest.rank['anotherElement']).toBe(1);
  });

  test('test union by rank', () => {
    forest.makeSet('1stElement');
    forest.makeSet('2ndElement');
    forest.makeSet('3rdElement');

    forest.union('2ndElement', '1stElement');

    // Instead of using 3rdElement as the new root node,
    // it uses 1stElement as the new root node,
    // because higher ranks will always win over dominantNode.
    forest.union('1stElement', '3rdElement');

    expect(forest.predecessor['1stElement']).toBe('1stElement');
    expect(forest.predecessor['2ndElement']).toBe('1stElement');
    expect(forest.predecessor['3rdElement']).toBe('1stElement');
  });

  test('test union is always merging root nodes', () => {
    forest.makeSet('1stElement');
    forest.makeSet('2ndElement');
    forest.makeSet('3rdElement');
    forest.makeSet('4thElement');

    expect(forest.predecessor['1stElement']).toBe('1stElement');
    expect(forest.predecessor['2ndElement']).toBe('2ndElement');
    expect(forest.predecessor['3rdElement']).toBe('3rdElement');
    expect(forest.predecessor['4thElement']).toBe('4thElement');

    forest.union('2ndElement', '1stElement');
    forest.union('4thElement', '3rdElement');

    expect(forest.predecessor['1stElement']).toBe('1stElement');
    expect(forest.predecessor['2ndElement']).toBe('1stElement');
    expect(forest.predecessor['3rdElement']).toBe('3rdElement');
    expect(forest.predecessor['4thElement']).toBe('3rdElement');

    forest.union('3rdElement', '1stElement');

    expect(forest.predecessor['1stElement']).toBe('1stElement');
    expect(forest.predecessor['2ndElement']).toBe('1stElement');
    expect(forest.predecessor['3rdElement']).toBe('1stElement');
    expect(forest.predecessor['4thElement']).toBe('3rdElement');
  });

  test('test findSet with path compression', () => {
    forest.makeSet('1stElement');
    forest.makeSet('2ndElement');
    forest.makeSet('3rdElement');

    forest.union('2ndElement', '1stElement');

    expect(forest.predecessor['3rdElement']).toBe('3rdElement');

    forest.union('3rdElement', '2ndElement');

    expect(forest.findSet('3rdElement')).toBe('1stElement');
    expect(forest.predecessor['3rdElement']).toBe('1stElement');
  });

  test('test findSet fails for unknown element', () => {
    expect(() => {
      forest.findSet('nonExistingElement');
    }).toThrow();
  });

  test('test toString provides correct representation', () => {
    forest.makeSet('someElement');
    forest.makeSet('anotherElement');

    forest.union('someElement', 'anotherElement');
    const expected =
      'Forest(predecessor={"someElement":"anotherElement","anotherElement":"anotherElement"}, rank={"someElement":0,"anotherElement":1})';
    expect(forest.toString()).toBe(expected);
  });
});
